src/x86_64/page.c

```diff
@@ -243,53 +294,65 @@ static inline boolean map_page(page base, u64 v, physical p, heap h,
     return true;
 }
 
-static inline u64 pt_level_end(u64 p, int level)
+//#define TRAVERSE_PTES_DEBUG
+
+#define PTE_ENTRIES U64_FROM_BIT(9)
+static boolean recurse_ptes(u64 pbase, int level, u64 vstart, u64 len, u64 laddr, entry_handler ph)
 {
-    return (p & ~MASK(level)) + U64_FROM_BIT(level);
-}
+    int shift = level_shift[level];
+    u64 lsize = U64_FROM_BIT(shift);
+    u64 start_idx = vstart > laddr ? ((vstart - laddr) >> shift) : 0;
+    u64 x = vstart + len - laddr;
+    u64 end_idx = MIN(pad(x, lsize) >> shift, PTE_ENTRIES);
+    u64 offset = start_idx << shift;
 
-#define for_level(base, start, end, level, levelend)                    \
-    for (u64 addr ## level = start, next ## level, end ## level, * pte ## level; \
-         next ## level = pt_level_end(addr ## level, PT ## level),      \
-             end ## level = MIN(next ## level, end),                    \
-             pte ## level = ((u64*)base) + pindex(addr ## level, PT ## level), \
-             addr ## level < levelend;                                  \
-         addr ## level = next ## level)
+#ifdef TRAVERSE_PTES_DEBUG
+    rprintf("   pbase 0x%lx, level %d, shift %d, lsize 0x%lx, laddr 0x%lx,\n"
+            "      start_idx %ld, end_idx %ld, x 0x%lx, offset 0x%lx\n",
+            __func__, pbase, level, shift, lsize, laddr,
+            start_idx, end_idx, offset);
```

Comment:
The format string and the other arguments don't match: `__func__` shouldn't be there, and `x` is missing.